core/src/main/java/org/apache/druid/timeline/VersionedIntervalTimeline.java

```diff
@@ -75,12 +79,11 @@
       Comparators.intervalsByStartThenEnd()
   );
   private final Map<Interval, TreeMap<VersionType, TimelineEntry>> allTimelineEntries = new HashMap<>();
+  private final AtomicInteger numObjects = new AtomicInteger();
```

Comment:
`VersionedIntervalTimeline` is not thread-safe. Should this be just an integer? I'm not sure how you're planning to use `VersionedIntervalTimeline`.

Reply:
Or, this variable is used only in `iterateAllObjects`. Should it return a collection? Or can we change the return type to Stream or Iterator?

Reply:
`VersionedIntervalTimeline` actually has to be thread-safe: see #7285.

Reply:
It's more convenient for users of the class if the method returns a `Collection` rather than `Iterable` or `Stream`: it makes possible both for-each iteration and stream pipeline call. Although currently this method is used in just one place with `forEach()` call, I don't see much value of over-specializing for this case now.

Reply:
As aforementioned, we don't have to have `numObjects` if `iterateAllObjects()` returns an `Iterable` or `Stream`. As you noted, the only user is calling `forEach()` on the result, it doesn't have to be a collection. As far as I remember, we agreed on simplifying unnecessarily generalized code.

Reply:
I actually discovered a use of numObjects in #7306, from where this PR is partially extracted: https://github.com/apache/incubator-druid/pull/7306/files#diff-b3571a010cbbb3cb648676a89936a51fR277